synapse/handlers/saml_handler.py

```diff
@@ -95,6 +101,181 @@ def handle_redirect_request(self, client_redirect_url):
         # this shouldn't happen!
         raise Exception("prepare_for_authenticate didn't return a Location header")
 
+    # User should have only one local session to only one entity
+    # But in case of crash we may have multiple sessions stored in the
+    # local cache. In that case we select the last session and remove
+    # other ones.
+    def _find_session_from_user(self, username):
+        subjects = self._saml_client.users.subjects()
```

Comment:
how does this work when the server is restarted?

Reply:
I think it's important that logout will continue to work for users who have been logged in for longer than the process has been running.

I guess we need to store the necessary information in the synapse database, against the user's device id.